### PR TITLE
JBIDE-22078 - full publish / clean publish broken on windows for local wf adapters

### PR DESCRIPTION
**Pull request by robstryker**

### CODE REVIEW

File: as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/StandardFileSystemPublishController.java [hunk 1]
```diff
@@ -432,7 +431,17 @@ private int handlePublishBinaryModule(IModule[] module, IPath archiveDestination
 		return IServer.PUBLISH_STATE_UNKNOWN;
 	}
 	
-	private int removeModule(IModule[] module, IPath remote, IProgressMonitor monitor) throws CoreException {
+	
+	protected IStatus[] executeFullPublish(IModule[] module, IPath archiveDestination, IModulePathFilter filter, IProgressMonitor monitor) 
+			throws CoreException {
+		PublishModuleFullRunner runner = new PublishModuleFullRunner(getFilesystemController(), archiveDestination);
+		IModuleResource[] filtered = filter == null ? ModuleResourceUtil.getMembers(module[module.length-1]) : filter.getFilteredMembers();
```
Comment (fbricon) [inline as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/StandardFileSystemPublishController.java]: can module be empty at this point?, cause module[-1] wouldn't work that well

File: as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.IServer;
+import org.jboss.ide.eclipse.as.core.server.IModulePathFilter;
+import org.jboss.ide.eclipse.as.core.server.internal.v7.DeploymentMarkerUtils;
+import org.jboss.ide.eclipse.as.wtp.core.server.behavior.LocalFilesystemController;
+
+public class WildFlyFileSystemPublishController extends StandardFileSystemPublishController {
```
Comment (fbricon) [inline as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java]: I wonder if it also happens on older (pre-WF) AS versions

File: as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+				getFilesystemController() instanceof LocalFilesystemController ) {
+			DeploymentMarkerUtils.removeDeployedMarker(remote, getFilesystemController());
+			long timeout = System.currentTimeMillis() + maxWait;
+			boolean done = false;
+			while( !done && System.currentTimeMillis() < timeout) {
+				// Now wait
+				done = DeploymentMarkerUtils.markerExists(remote, getFilesystemController(), DeploymentMarkerUtils.UNDEPLOYED);
+				if( !done ) {
+					try {
+						Thread.sleep(500);
```
Comment (fbricon) [inline as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java]: half a second is pretty long. Maybe wait 200ms? Depends on the cost of calling DeploymentMarkerUtils.markerExists I guess.

File: as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import org.jboss.ide.eclipse.as.wtp.core.server.behavior.LocalFilesystemController;
+
+public class WildFlyFileSystemPublishController extends StandardFileSystemPublishController {
+	protected int removeModule(IModule[] module, IPath remote, IProgressMonitor monitor) throws CoreException {
+		undeployModuleAndWait(module, remote, 15000, monitor);
+		return super.removeModule(module, remote, monitor);
+	}
+	protected IStatus[] executeFullPublish(IModule[] module, IPath archiveDestination, IModulePathFilter filter, IProgressMonitor monitor) 
+			throws CoreException {
+		undeployModuleAndWait(module, archiveDestination, 15000, monitor);
```
Comment (fbricon) [inline as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java]: should we proceed if undeployModuleAndWait return false? at the very least we should log the failed un-deployment attempt

File: as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/StandardFileSystemPublishController.java (hunk 1 above)
Comment (robstryker) [inline as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/StandardFileSystemPublishController.java]: impossible

File: as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java (hunk 2 above)
Comment (robstryker) [inline as/plugins/org.jboss.ide.eclipse.as.core/jbosscore/org/jboss/tools/as/core/server/controllable/subsystems/internal/WildFlyFileSystemPublishController.java]: as6 and below do not make use of the same filewatcher abilities, and so the odds of such locks are very very very small. Also, our code for as6 and below actually suspends the deployment scanners before changing any files.

**Comment by fbricon** [comment]
Tested on Windows 10 with EAP 6.4, EAP 7, WF 8 and WF 10. @robstryker fixed windows :-)
Big +1 to apply

@alexeykazakov, your move!

**Comment by alexeykazakov** [comment]
+1

**Pull request merged**